File: utilities/src/main/java/com/wilsofts/utilities/network/download/RetrofitDownloader.java
package com.wilsofts.utilities.network.download;

import android.os.AsyncTask;

import androidx.annotation.NonNull;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class RetrofitDownloader extends AsyncTask<ResponseBody, Long, String> {
    private final DownloadResponse downloadResponse;
    private final File target_path;

    public RetrofitDownloader(Call<ResponseBody> call, File target_path, DownloadResponse downloadResponse) {
        this(call, downloadResponse, target_path);
    }

    public RetrofitDownloader(Call<ResponseBody> call, DownloadResponse downloadResponse, File target_path) {
        this.downloadResponse = downloadResponse;
        this.target_path = target_path;

        call.enqueue(new Callback<ResponseBody>() {
            @Override
            public void onResponse(@NonNull Call<ResponseBody> call, @NonNull final Response<ResponseBody> response) {
                if (response.isSuccessful()) {
                    RetrofitDownloader.this.execute(response.body());
                } else {
                    RetrofitDownloader.this.downloadResponse.response(response.code());
                }
            }

            @Override
            public void onFailure(@NonNull Call<ResponseBody> call, @NonNull Throwable throwable) {
                RetrofitDownloader.this.downloadResponse.error(throwable);
            }
        });
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();

    }

    @Override
    protected String doInBackground(ResponseBody... urls) {
        //Copy you logic to calculate progress and call
        this.saveToDisk(urls[0]);
        return null;
    }

    @Override
    protected void onProgressUpdate(Long... progress) {
        super.onProgressUpdate(progress);
        this.downloadResponse.progress(progress[0], progress[1]);
    }

    @Override
    protected void onPostExecute(String result) {

    }

    private void saveToDisk(ResponseBody body) {
        InputStream inputStream = null;
        OutputStream outputStream = null;

        try {
            inputStream = body.byteStream();
            outputStream = new FileOutputStream(this.target_path);
            byte[] data = new byte[4096];
            int count;
            long progress = 0;
            long fileSize = body.contentLength();

            while ((count = inputStream.read(data)) != -1) {
                outputStream.write(data, 0, count);
                progress += count;
                this.publishProgress(progress, fileSize);
            }

            outputStream.flush();
            this.publishProgress(fileSize, fileSize);

        } catch (IOException error) {
            this.publishProgress(-1L, -1L);
            this.downloadResponse.error(error);

        } finally {
            try {
                if (inputStream != null) {
                    inputStream.close();
                }
                if (outputStream != null) {
                    outputStream.close();
                }
            } catch (IOException error) {
                error.printStackTrace();
            }
        }
    }
}

File: utilities/src/main/java/com/wilsofts/utilities/network/download/RetrofitDownloader.kt
package com.wilsofts.utilities.network.download

import android.os.AsyncTask
import com.wilsofts.utilities.LibUtils
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.*

class RetrofitDownloader(call: Call<ResponseBody>, private val downloadResponse: DownloadResponse, private val target_path: File)
    : AsyncTask<ResponseBody, RetrofitDownloader.Progress, String>() {

    constructor(call: Call<ResponseBody>, target_path: File, downloadResponse: DownloadResponse) : this(call, downloadResponse, target_path)

    init {
        call.enqueue(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    this@RetrofitDownloader.execute(response.body())
                } else {
                    this@RetrofitDownloader.downloadResponse.response(response.code())
                }
            }

            override fun onFailure(call: Call<ResponseBody>, throwable: Throwable) {
                this@RetrofitDownloader.downloadResponse.error(throwable)
            }
        })
    }

    override fun doInBackground(vararg urls: ResponseBody): String? {
        //Copy you logic to calculate progress and call
        this.saveToDisk(urls[0])
        return null
    }

    override fun onProgressUpdate(vararg values: Progress?) {
        super.onProgressUpdate(*values)
        this.downloadResponse.progress(values[0]!!.progress, values[0]!!.file_size)
    }

    override fun onPostExecute(result: String) {

    }

    private fun saveToDisk(body: ResponseBody) {
        var inputStream: InputStream? = null
        var outputStream: OutputStream? = null

        try {
            inputStream = body.byteStream()
            outputStream = FileOutputStream(this.target_path)
            val data = ByteArray(4096)
            var count: Int = inputStream!!.read(data)
            var progress: Long = 0
            val fileSize = body.contentLength()

            while ((count) != -1) {
                outputStream.write(data, 0, count)
                progress += count.toLong()
                this.publishProgress(Progress(progress, fileSize))
                count = inputStream.read(data)
            }

            outputStream.flush()
            this.publishProgress(Progress(fileSize, fileSize))

        } catch (error: IOException) {
            this.publishProgress(Progress(-1L, -1L))
            this.downloadResponse.error(error)

        } finally {
            try {
                inputStream?.close()
                outputStream?.close()
            } catch (error: IOException) {
                LibUtils.logE(error)
            }
        }
    }

    class Progress(val progress: Long, val file_size: Long)
}
